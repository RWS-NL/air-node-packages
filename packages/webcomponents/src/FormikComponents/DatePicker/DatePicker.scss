@import "../../styles/vars";

.outline {
  border-radius: 0 !important;
  border-width: 1px;
}

.input {
  color: $textBlack;
  font-size: 0.9rem;
  height: 2.5rem;
  min-height: 2.5rem;

  &Label {
    font-size: 1rem;
    font-style: italic;
    transform: translate(14px, -6px) scale(0.75);
  }
}

.error {
  &Label {
    position: absolute;
    top: 2.5rem;
  }
}

.datePickerIcon {
  button {
    padding: 0.4rem;

    &:hover {
      padding: 0.4rem;
    }
  }

  svg {
    height: 1rem;
    width: auto;
  }
}
