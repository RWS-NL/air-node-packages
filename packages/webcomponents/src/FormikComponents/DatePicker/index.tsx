import { KeyboardDatePicker, KeyboardDatePickerProps } from '@material-ui/pickers';
import { MaterialUiPickersDate } from '@material-ui/pickers/typings/date';
import { FieldProps } from 'formik';
import React, { FC } from 'react';
import css from './DatePicker.scss';

type DatePickerFormProps = DatePickerProps &
  FieldProps &
  Omit<KeyboardDatePickerProps, 'error' | 'name' | 'onChange' | 'value' | 'variant'>;

export interface DatePickerProps {
  /** The variant of the datepicker */
  variant: 'standard' | 'filled' | 'outlined' | undefined;
  /** The label shown in the datepicker */
  label: string;
  /** The minimum date to meet */
  minDate: Date;
}

/**
 * Constructs a Date Picker with formik validation
 * @param componentProps Props to pass to the date picker component
 * @example
 * ```jsx
 * <FastField
 *   name='startDate'
 *   label='Project startdatum'
 *   minDate={new Date('2012-01-01')}
 *   component={DatePicker}
 * />
 * ```
 */
export const DatePicker: FC<DatePickerProps> = componentProps => {
  const { field, form, label, minDate, ...props } = componentProps as DatePickerFormProps;

  const currentError = form.errors[field.name];

  return (
    <KeyboardDatePicker
      {...props}
      variant='inline'
      format='DD/MM/YYYY'
      minDate={minDate}
      name={field.name}
      onBlur={field.onBlur}
      value={field.value}
      helperText={currentError}
      error={Boolean(currentError)}
      onChange={(newDate: MaterialUiPickersDate | null) =>
        form.setFieldValue(field.name, newDate?.toDate() || new Date(), true)
      }
      KeyboardButtonProps={{ 'aria-label': 'change date' }}
      InputAdornmentProps={{ position: 'start', className: css.datePickerIcon }}
      label={label}
      inputVariant='outlined'
      fullWidth
      autoOk
      animateYearScrolling
      InputProps={{
        classes: {
          root: css.input,
          notchedOutline: css.outline
        }
      }}
      InputLabelProps={{
        shrink: true,
        classes: { root: css.inputLabel }
      }}
      FormHelperTextProps={{
        component: 'div',
        classes: { error: css.errorLabel }
      }}
    />
  );
};

export default DatePicker;
