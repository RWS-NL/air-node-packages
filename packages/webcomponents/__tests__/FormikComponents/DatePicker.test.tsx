import DatePicker from '../../src/FormikComponents/DatePicker/DatePicker';
import { shallow } from 'enzyme';
import React from 'react';

const minimumDateForPicker = new Date('2012-01-01');
const mockOnchange = jest.fn();
const mockOnBlur = jest.fn();
const mockForm = { errors: {}, setFieldValue: mockOnchange };
const mockField = { name: '', onBlur: mockOnBlur };

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    const datePicker = shallow(
      <DatePicker variant='standard' label='testDatePicker' minDate={minimumDateForPicker}
        // @ts-ignore:2322 => Required because TS doesn't know these props can be passed from Formik
        form={mockForm} field={mockField}
      />
    );
    expect(datePicker).toMatchSnapshot();
  });
});