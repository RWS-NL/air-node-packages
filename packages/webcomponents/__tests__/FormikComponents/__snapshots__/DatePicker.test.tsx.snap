// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Required Props 1`] = `
<PickerWithState
  FormHelperTextProps={
    Object {
      "classes": Object {
        "error": "errorLabel",
      },
      "component": "div",
    }
  }
  InputAdornmentProps={
    Object {
      "className": "datePickerIcon",
      "position": "start",
    }
  }
  InputLabelProps={
    Object {
      "classes": Object {
        "root": "inputLabel",
      },
      "shrink": true,
    }
  }
  InputProps={
    Object {
      "classes": Object {
        "notchedOutline": "outline",
        "root": "input",
      },
    }
  }
  KeyboardButtonProps={
    Object {
      "aria-label": "change date",
    }
  }
  allowKeyboardControl={true}
  animateYearScrolling={true}
  autoOk={true}
  error={false}
  format="DD/MM/YYYY"
  fullWidth={true}
  inputVariant="outlined"
  invalidDateMessage="Invalid Date Format"
  label="testDatePicker"
  maxDate={2100-01-01T00:00:00.000Z}
  maxDateMessage="Date should not be after maximal date"
  minDate={2012-01-01T00:00:00.000Z}
  minDateMessage="Date should not be before minimal date"
  name=""
  onBlur={[MockFunction]}
  onChange={[Function]}
  openTo="date"
  variant="inline"
  views={
    Array [
      "year",
      "date",
    ]
  }
/>
`;
